src/mlpack/core/tree/spill_tree/spill_dual_tree_traverser_impl.hpp

```diff
@@ -113,8 +113,18 @@ SpillDualTreeTraverser<RuleType, Defeatist>::Traverse(
       size_t bestChild = rule.GetBestChild(queryNode, referenceNode);
       if (bestChild < referenceNode.NumChildren())
       {
-        Traverse(queryNode, referenceNode.Child(bestChild));
-        ++numPrunes;
+        // If the bestchild node has less number of points than the minimum
+        // basecases to be run, recurse into other node as well.
+        if (referenceNode.Child(bestChild).NumPoints() <= rule.MinBaseCases())
+        {
+          Traverse(queryNode, referenceNode.Child(bestChild));
+          Traverse(queryNode, referenceNode.Child(!bestChild));
```

Comment:
Oh, I think that the issue might be here.  Try replacing this `Traverse()` instead with a loop over all the descendants, like this:

```
for (size_t q = 0; q < queryNode.NumDescendants(); ++q)
  for (size_t r = 0; r < referenceNode.NumDescendants(); ++r)
    rule.BaseCase(queryNode.Descendant(q), referenceNode.Descendant(r));

numBaseCases += queryNode.NumDescendants() * referenceNode.NumDescendants();
++numPrunes;
return;
```

Basically the idea is to avoid traversing into either reference child, since if we traverse into *both* then we might call `BaseCase()` with a reference point that's in *both* the left and right reference child.  But if we do the base cases at this level only using `Descendant()`, we now know (thanks to the patch I gave) that we will not get any duplicate reference points like that.

I think the idea I gave above is not perfect, so you will probably have to adapt it a little bit, but I hope it gets across the basic idea.  Let me know if not and I can clarify. :+1: